### PR TITLE
IO: Refine the `CompressedReadBuffer`/`CompressedWriteBuffer`

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/8691

Problem Summary:
`CompressedReadBufferFromFileProvider<has_checksum=false>` is misleading. 

### What is changed and how it works?

[Done] <del> Should merge https://github.com/pingcap/tiflash/pull/8757 first </del>
Refine the code by providing some static methods and aliases. Also added some comments about the background.
A sad fact is that there are lots of code rely on the legacy checksum. The FramedChecksum only applied to DMFile.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Encryption/CompressedReadBufferFromFileProvider.cpp [hunk 1]
```diff
@@ -25,8 +25,8 @@ namespace ErrorCodes
 extern const int SEEK_POSITION_OUT_OF_BOUND;
 }
 
-template <bool has_checksum>
-bool CompressedReadBufferFromFileProvider<has_checksum>::nextImpl()
+template <bool has_legacy_checksum>
+bool CompressedReadBufferFromFileProviderImpl<has_legacy_checksum>::nextImpl()
```
Comment (Lloyd-Pottiger) [inline dbms/src/Encryption/CompressedReadBufferFromFileProvider.cpp:29]: Maybe move to `IO/` after https://github.com/pingcap/tiflash/pull/8763 merged.

File: dbms/src/IO/Compression/CompressedWriteBuffer.h [hunk 2]
```diff
@@ -41,6 +48,18 @@ class CompressedWriteBuffer : public BufferWithOwnMemory<WriteBuffer>
         CompressionSettings compression_settings = CompressionSettings(),
         size_t buf_size = DBMS_DEFAULT_BUFFER_SIZE);
 
+    static std::unique_ptr<WriteBuffer> build(
```
Comment (Lloyd-Pottiger) [inline dbms/src/IO/Compression/CompressedWriteBuffer.h:51]:
```suggestion
    static std::unique_ptr<CompressedWriteBuffer<true>> build(
```

return `std::unique_ptr<CompressedWriteBuffer<true>>` ?

Comment (JaySon-Huang) [reply dbms/src/IO/Compression/CompressedWriteBuffer.h:51]: When `enable_legacy_checksum` is false, this function return `std::unique_ptr<CompressedWriteBuffer<false>>`, which can not be hold by `std::unique_ptr<CompressedWriteBuffer<true>>`

Comment (Lloyd-Pottiger) [reply dbms/src/IO/Compression/CompressedWriteBuffer.h:51]: ok

**Review by Lloyd-Pottiger — APPROVED** [review]

**Pull request merged**